pennylane_lightning/lightning_qubit/_state_vector.py

```diff
@@ -0,0 +1,386 @@
+            try:
+                method(
+                    qml.matrix(operation.base),
+                    control_wires,
+                    control_values,
+                    target_wires,
+                    False,
+                )
+            except AttributeError:  # pragma: no cover
+                # To support older versions of PL
+                method(operation.base.matrix, control_wires, control_values, target_wires, False)
+
+    def apply_lightning(self, operations):
+        """Apply a list of operations to the state tensor.
+
+        Args:
+            operations (list[~pennylane.operation.Operation]): operations to apply
+
+        Returns:
+            array[complex]: the output state tensor
+        """
+        state = self.state_vector
+
+        # Skip over identity operations instead of performing
+        # matrix multiplication with it.
+        for operation in operations:
+            name = operation.name
+            if name == "Identity":
+                continue
+            method = getattr(state, name, None)
+            wires = self.wires.indices(operation.wires)
+
+            if method is not None:  # apply specialized gate
+                inv = False
+                param = operation.parameters
+                method(wires, inv, param)
+            elif (
+                name[0:2] == "C(" or name == "ControlledQubitUnitary" or name == "MultiControlledX"
+            ):  # apply n-controlled gate
+                print("hi")
+                self._apply_lightning_controlled(operation)
+            else:  # apply gate as a matrix
+                # Inverse can be set to False since qml.matrix(operation) is already in
+                # inverted form
+                method = getattr(state, "applyMatrix")
+                try:
+                    method(qml.matrix(operation), wires, False)
+                except AttributeError:  # pragma: no cover
+                    # To support older versions of PL
+                    method(operation.matrix, wires, False)
+
+    def apply_operations(self, operations):
+        """Applies operations to the state vector."""
+        # State preparation is currently done in Python
+        if operations:  # make sure operations[0] exists
+            if isinstance(operations[0], StatePrep):
+                self._apply_state_vector(operations[0].parameters[0].copy(), operations[0].wires)
+                operations = operations[1:]
+            elif isinstance(operations[0], BasisState):
+                self._apply_basis_state(operations[0].parameters[0], operations[0].wires)
+                operations = operations[1:]
+
+        for operation in operations:
+            if isinstance(operation, (StatePrep, BasisState)):
+                raise DeviceError(
+                    f"Operation {operation.name} cannot be used after other "
+                    f"Operations have already been applied on a {self.short_name} device."
+                )
```

Comment:
Should be handled by preprocessing.  Probably fine to leave in, but we could also remove it. If someone skipped calling preprocessing, they are asking for things to fall over in confusing ways.

Reply:
I will remove it.